`esp32s3-hal` can no longer be built with the `direct-boot` feature enabled

When direct boot support was initially added in 9fa1d1ecdbb1621657971e0ad9be138a015a2757 it was working, but it seems that this changed in (or before?) 22ac479bff5fd52b8d2ee984b1479eca9b981930.

When trying to build I get a linker error:

```text
error: linking with `xtensa-esp32s3-elf-gcc` failed: exit status: 1
  |
  = note: "xtensa-esp32s3-elf-gcc" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/examples/blinky-4a71c994db6341dd.blinky.9e4138a0-cgu.0.rcgu.o" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/examples/blinky-4a71c994db6341dd.blinky.9e4138a0-cgu.1.rcgu.o" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/examples/blinky-4a71c994db6341dd.blinky.9e4138a0-cgu.2.rcgu.o" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/examples/blinky-4a71c994db6341dd.blinky.9e4138a0-cgu.3.rcgu.o" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/examples/blinky-4a71c994db6341dd.blinky.9e4138a0-cgu.4.rcgu.o" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/examples/blinky-4a71c994db6341dd.blinky.9e4138a0-cgu.5.rcgu.o" "-Wl,--as-needed" "-L" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps" "-L" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/release/deps" "-L" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/build/esp32s3-hal-719ba7df5f143918/out" "-L" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/build/xtensa-lx-rt-8927f707e7183454/out" "-L" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/build/esp32s3-2a1e2f5103e19372/out" "-L" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/build/xtensa-lx-33fa9597464e9016/out" "-L" "/Users/jesse/.rustup/toolchains/esp/lib/rustlib/xtensa-esp32s3-none-elf/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp_backtrace-76d61018c671f9af.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp_println-07e8828d4267fc4a.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp32s3_hal-3b05d96ed27c9a81.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp_hal_common-cee87ed226eeae7f.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libcritical_section-f1833d30027e4bd6.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libcfg_if-acade488a4183f4a.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libxtensa_lx_rt-c30a739f6f02924a.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libr0-5e29015152a5db6a.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libfugit-71c3058bcbd2d2f3.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libgcd-eab82d1c0cb14b3a.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp32s3-7c6f25e9cd8ec492.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libxtensa_lx-828d0d8482c072c6.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libspin-16c56e32c76f5ec8.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/liblock_api-8c664f495cf01cba.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libscopeguard-fc26c31819741e13.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libmutex_trait-d25535b6040a2417.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libbare_metal-1d83263e925fd7d9.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libvcell-dbd9595000d3818d.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libembedded_hal-788a104818511a30.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libvoid-9030f8c26cccf36a.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libnb-be252d4bdbd89b07.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libnb-3774a76e7e5de9c6.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/librustc_std_workspace_core-8982f2ea24a0a7d8.rlib" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libcore-7a8dc6e6fc0a63de.rlib" "-Wl,--end-group" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libcompiler_builtins-a08b3a2d4f692a3b.rlib" "-Wl,-Bdynamic" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/Users/jesse/.rustup/toolchains/esp/lib/rustlib/xtensa-esp32s3-none-elf/lib" "-o" "/Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/examples/blinky-4a71c994db6341dd" "-Wl,--gc-sections" "-no-pie" "-Wl,-O1" "-nodefaultlibs" "-nostartfiles" "-Wl,-Tlinkall.x"
  = note: /Users/jesse/.espressif/tools/xtensa-esp32s3-elf/esp-2022r1-11.2.0/xtensa-esp32s3-elf/bin/../lib/gcc/xtensa-esp32s3-elf/11.2.0/../../../../xtensa-esp32s3-elf/bin/ld: /Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp_hal_common-cee87ed226eeae7f.rlib(esp_hal_common-cee87ed226eeae7f.esp_hal_common.5b8be18a-cgu.3.rcgu.o):(.rwtext.literal+0x28): undefined reference to `Timer0'
          /Users/jesse/.espressif/tools/xtensa-esp32s3-elf/esp-2022r1-11.2.0/xtensa-esp32s3-elf/bin/../lib/gcc/xtensa-esp32s3-elf/11.2.0/../../../../xtensa-esp32s3-elf/bin/ld: /Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp_hal_common-cee87ed226eeae7f.rlib(esp_hal_common-cee87ed226eeae7f.esp_hal_common.5b8be18a-cgu.3.rcgu.o):(.rwtext.literal+0x30): undefined reference to `Software1'
          /Users/jesse/.espressif/tools/xtensa-esp32s3-elf/esp-2022r1-11.2.0/xtensa-esp32s3-elf/bin/../lib/gcc/xtensa-esp32s3-elf/11.2.0/../../../../xtensa-esp32s3-elf/bin/ld: /Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp_hal_common-cee87ed226eeae7f.rlib(esp_hal_common-cee87ed226eeae7f.esp_hal_common.5b8be18a-cgu.3.rcgu.o):(.rwtext.literal+0x34): undefined reference to `Timer2'
          /Users/jesse/.espressif/tools/xtensa-esp32s3-elf/esp-2022r1-11.2.0/xtensa-esp32s3-elf/bin/../lib/gcc/xtensa-esp32s3-elf/11.2.0/../../../../xtensa-esp32s3-elf/bin/ld: /Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp_hal_common-cee87ed226eeae7f.rlib(esp_hal_common-cee87ed226eeae7f.esp_hal_common.5b8be18a-cgu.3.rcgu.o):(.rwtext.literal+0x38): undefined reference to `Timer1'
          /Users/jesse/.espressif/tools/xtensa-esp32s3-elf/esp-2022r1-11.2.0/xtensa-esp32s3-elf/bin/../lib/gcc/xtensa-esp32s3-elf/11.2.0/../../../../xtensa-esp32s3-elf/bin/ld: /Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp_hal_common-cee87ed226eeae7f.rlib(esp_hal_common-cee87ed226eeae7f.esp_hal_common.5b8be18a-cgu.3.rcgu.o):(.rwtext.literal+0x3c): undefined reference to `NMI'
          /Users/jesse/.espressif/tools/xtensa-esp32s3-elf/esp-2022r1-11.2.0/xtensa-esp32s3-elf/bin/../lib/gcc/xtensa-esp32s3-elf/11.2.0/../../../../xtensa-esp32s3-elf/bin/ld: /Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp_hal_common-cee87ed226eeae7f.rlib(esp_hal_common-cee87ed226eeae7f.esp_hal_common.5b8be18a-cgu.3.rcgu.o):(.rwtext.literal+0x40): undefined reference to `Profiling'
          /Users/jesse/.espressif/tools/xtensa-esp32s3-elf/esp-2022r1-11.2.0/xtensa-esp32s3-elf/bin/../lib/gcc/xtensa-esp32s3-elf/11.2.0/../../../../xtensa-esp32s3-elf/bin/ld: /Users/jesse/Work/esp-hal/esp32s3-hal/target/xtensa-esp32s3-none-elf/release/deps/libesp_hal_common-cee87ed226eeae7f.rlib(esp_hal_common-cee87ed226eeae7f.esp_hal_common.5b8be18a-cgu.3.rcgu.o):(.rwtext.literal+0x44): undefined reference to `Software0'
          collect2: error: ld returned 1 exit status

  = help: some `extern` functions couldn't be found; some native libraries may need to be installed or have their path specified
  = note: use the `-l` flag to specify native libraries to link
  = note: use the `cargo:rustc-link-lib` directive to specify the native libraries to link with Cargo (see https://doc.rust-lang.org/cargo/reference/build-scripts.html#cargorustc-link-libkindname)

error: could not compile `esp32s3-hal` due to previous error
```

I suppose we should check this feature in CI for the supported chips as well.

--- a/esp32s3-hal/ld/db-esp32s3.x
+++ b/esp32s3-hal/ld/db-esp32s3.x
@@ -58,6 +58,17 @@ EXTERN(__default_exception);
 EXTERN(__default_double_exception);
 EXTERN(__default_interrupt);
 
+/* high level CPU interrupts */
+PROVIDE(Timer0 = __default_user_exception);
+PROVIDE(Timer1 = __default_user_exception);
+PROVIDE(Timer2 = __default_user_exception);
+PROVIDE(Timer3 = __default_user_exception);
+PROVIDE(Profiling = __default_user_exception);
+PROVIDE(NMI = __default_user_exception);
+PROVIDE(Software0 = __default_user_exception);
+PROVIDE(Software1 = __default_user_exception);
+
+/* low level exception/interrupt, which must be overridden using naked functions */
 EXTERN(__default_naked_exception);
 EXTERN(__default_naked_double_exception);
 EXTERN(__default_naked_level_2_interrupt);

